src/communication/communicator_manager.cpp

```diff
@@ -82,15 +85,55 @@ class communicator_manager::implementation
         return result;
     }
 
-    std::shared_ptr<communicator>
-    get_world_communicator(const std::string &name) const
-    {
-        std::shared_ptr<communicator> result;
-        
-        const auto* backend = get_backend(name);
-        if (backend)
+    communicator_backend* get_preferred_backend() const
+    {   
+        communicator_backend* result;
+
+        // Filter available backends
+        std::vector<communicator_backend*> available_backends;
+        available_backends.reserve(m_registry.size());
+        for (auto &item : m_registry)
+        {   
+            if (item.second->is_available())
+            {
+                available_backends.emplace_back(item.second.get());
+            }
+        }
+
+        if( available_backends.empty())
+        {
+            result = nullptr;
+        }
+        else if (available_backends.size() == 1)
         {
-            result = backend->get_world_communicator();
+            result = available_backends[0];
+        }
+        else // available_backends.size() > 1
+        {
+            // Get the best two
+            std::partial_sort(
+                available_backends.begin(),
+                std::next(available_backends.begin(), 2),
+                available_backends.end(),
+                [] (const communicator_backend *lhs, 
+                    const communicator_backend *rhs )
+                {
+                    return lhs->get_priority() > rhs->get_priority();
+                }
+            );
+
+            // Ensure the second is worse
+            if (available_backends[0]->get_priority() == 
+                available_backends[1]->get_priority() )
+            {
+                throw ambiguous_backend_error(
+                    "Could not disambiguate multiple among multiple "
```

Comment:
`Could not disambiguate among multiple ...` (I think there is one extra "multiple")